lib/common/crmcommon_private.h

```diff
@@ -428,12 +428,16 @@ enum pcmk__schema_validator {
 typedef struct {
     int schema_index;
     char *name;
-    char *transform;
+
+    /*!
+     * List of XSLT stylesheets for upgrading from this schema version to the
+     * next one. Sorted by the order in which they should be applied to the CIB.
+     */
+    GList *transforms;
+
```

Comment:
What's the rationale for supporting arbitrary transforms? The original model requires a single transform, and allows a pair of optional enter/leave transforms. This loses the optional vs required aspect and the idea that one of the transforms is the "main" one that does most of the work.

Reply:
> rationale for supporting arbitrary transforms

It simplifies certain transformations that we want. As one example, one thing I'm working on is to drop upstart-class resources. This means:
* dropping templates with class=upstart
* dropping primitives with class=upstart or referencing a template with class=upstart
* dropping groups that are empty after removing these primitives
* dropping clones that are empty after removing these primitives and groups
* dropping constraints that reference resources that no longer exist after these removals

XSL is declarative. One major implication is that the processing in stylesheets happens conceptually in parallel. So a single stylesheet is not meant to do transformations where one template's action depends on the result of applying some other template.

Another important property of XSL is closure. Input and output are both XML trees. This makes it easy to chain multiple stylesheets together into a pipeline, and that is a pretty standard and expected design pattern from what I've seen in my reading so far.

Transformations that involve a bunch of redundancy or are hard to read in a single stylesheet, become incredibly simple if we chain a few stylesheets together.

There may be a way to do the above with an XPath/variable trick but I haven't found one yet without using extensions. In any case, there may be other use cases for XSL pipelines as we continue to work through the transformations we need.

---

> The original model requires a single transform, and allows a pair of optional enter/leave transforms. This loses the optional vs required aspect and the idea that one of the transforms is the "main" one that does most of the work.

Correct, and I think the change is for the better. There's nothing magic about the number 3 or about having a "main" transformation. We've only used that pattern once (for 2.10 to 3.0). It worked because the transformations were architected so that enter and leave were symmetrical and hardly did anything (they mostly processed ID refs using attributes that were mangled versions of the originals). But look at `upgrade-2.10.xsl`. It's over 2500 lines and requires substantial effort to make sense of.

The smaller and more modular we can keep each of our templates, the easier this will all be to reason about.

Reply:
> There may be a way to do the above with an XPath/variable trick but I haven't found one yet without using extensions. In any case, there may be other use cases for XSL pipelines as we continue to work through the transformations we need.

I think I have found a way to do this using variables and an XPath 1.0 subset idiom (`count(X|Y) = count(Y)`). It is still much longer and uglier than it would be in a pipeline format, but it is workable.

Reply:
I think pipelines do make sense